api/catalog/api/controllers/elasticsearch/utils.py

```diff
@@ -0,0 +1,160 @@
+from itertools import accumulate
+from math import ceil
+from typing import List, Optional, Tuple
+
+from django.core.cache import cache
+
+from elasticsearch_dsl import Search
+from elasticsearch_dsl.response import Hit, Response
+
+from catalog.api.models import ContentProvider
+from catalog.api.utils.dead_link_mask import get_query_hash, get_query_mask
+from catalog.api.utils.validate_images import validate_images
+
+
+FILTER_CACHE_TIMEOUT = 30
+DEAD_LINK_RATIO = 1 / 2
+ELASTICSEARCH_MAX_RESULT_WINDOW = 10000
+
+
+def _exclude_filtered_providers(s: Search) -> Search:
+    """
+    Hide data sources from the catalog dynamically. This excludes providers with
+    ``filter_content`` enabled from the search results.
+
+    :param s: the search query to issue to Elasticsearch
+    :return: the modified search query
+    """
+
+    filter_cache_key = "filtered_providers"
+    filtered_providers = cache.get(key=filter_cache_key)
+    if not filtered_providers:
+        filtered_providers = ContentProvider.objects.filter(filter_content=True).values(
+            "provider_identifier"
+        )
+        cache.set(
+            key=filter_cache_key,
+            timeout=FILTER_CACHE_TIMEOUT,
+            value=filtered_providers,
+        )
```

Comment:
If the cache doesn't have any filtered providers, and no `ContentProvider`s have the field `filter_content=True`, it seems like this will attempt to update the cache every time the function is called. Is there a way to differentiate between "the cache doesn't have this value" and "this value is an empty list"?